### PR TITLE
ASC-247 avoid clean up of subspaces

### PR DESCRIPTION
**Pull request by mfrancisc**
This PR updates the space_cleanup_controller so that it doesn't sub-spaces without SpaceBindings.
SpaceBindings for sub-spaces will be inherited from the parent-space. 

It's covering following acceptance criteria from Jira https://issues.redhat.com/browse/ASC-247:

The cleanup space controller is updated so it cleans the Spaces up only if:
- there is no SpaceBinding for the given Space present
- at the same time, it either doesn't have parentSpace set or the parentSpace is gone or being deleted

Paired with:
- toolchain-e2e: https://github.com/codeready-toolchain/toolchain-e2e/pull/649

### CODE REVIEW

File: controllers/spacecleanup/space_cleanup_controller.go [hunk 1]
```diff
@@ -92,6 +92,11 @@ func (r *Reconciler) ensureDeletionIfNeeded(logger logr.Logger, space *toolchain
 		return false, 0, nil
 	}
 
+	// check if space has an "active" parentSpace
+	if exists, err := r.parentSpaceExists(logger, space); exists {
```
Comment (xcoulon) [inline controllers/spacecleanup/space_cleanup_controller.go:96]: what happens if `!exists && err != nil` ?

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 2]
```diff
@@ -104,6 +104,69 @@ func TestCleanupSpace(t *testing.T) {
 		assert.False(t, res.Requeue)
 	})
 
+	t.Run("with ParentSpace - Space shouldn't be deleted", func(t *testing.T) {
+		// given
+		parentSpace := spacetest.NewSpace("parentSpace",
+			spacetest.WithCreationTimestamp(time.Now().Add(-time.Minute)),
+		)
+		spaceBinding := spacebinding.NewSpaceBinding("johny", parentSpace.Name, "admin", "a-creator")
```
Comment (xcoulon) [inline controllers/spacecleanup/space_cleanup_controller_test.go:112]:
```suggestion
		parentSpaceBinding := spacebinding.NewSpaceBinding("johny", parentSpace.Name, "admin", "a-creator")
```
so the name of this binding clearly refers to the parent space

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 3]
```diff
@@ -104,6 +104,69 @@ func TestCleanupSpace(t *testing.T) {
...
+			Exists()
+	})
+
+	t.Run("with terminating ParentSpace - Space should be deleted", func(t *testing.T) {
+		// given
+		parentSpace := spacetest.NewSpace("parentSpace",
+			spacetest.WithCreationTimestamp(time.Now().Add(-time.Minute)),
+			spacetest.WithDeletionTimestamp(),
+		)
+		spaceBinding := spacebinding.NewSpaceBinding("johny", parentSpace.Name, "admin", "a-creator")
```
Comment (xcoulon) [inline controllers/spacecleanup/space_cleanup_controller_test.go:135]: same here: let's call it `parentSpaceBinding`

File: controllers/spacecleanup/space_cleanup_controller.go (hunk 1 above)
Comment (mfrancisc) [reply controllers/spacecleanup/space_cleanup_controller.go]: Thanks @xcoulon ! I've added a return of the err even in that case! 

See: [7eda771](https://github.com/codeready-toolchain/host-operator/pull/728/commits/7eda771d353d09901d0c50f121026a686e761946)

File: controllers/spacecleanup/space_cleanup_controller_test.go (hunk 2 above)
Comment (mfrancisc) [reply controllers/spacecleanup/space_cleanup_controller_test.go]: Changed in: [7eda771](https://github.com/codeready-toolchain/host-operator/pull/728/commits/7eda771d353d09901d0c50f121026a686e761946)

File: controllers/spacecleanup/space_cleanup_controller_test.go (hunk 3 above)
Comment (mfrancisc) [reply controllers/spacecleanup/space_cleanup_controller_test.go]: Changed in: [7eda771](https://github.com/codeready-toolchain/host-operator/pull/728/commits/7eda771d353d09901d0c50f121026a686e761946)

**Review by xcoulon — APPROVED** [review]
thanks for taking my comments into account @mfrancisc ;)

**Review by MatousJobanek — APPROVED** [review]

**Pull request merged**